worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,1062 @@
+					continue
+				}
+			} else {
+				self.shards.push(*current_shard);
+				new_shard = true;
+			}
+			// add block to DB batch
+			self.add_last_block(&mut batch, &signed_block);
+		}
+		// update stored_shards_key -> vec<shard> only if a new shard was included
+		if new_shard {
+			SidechainDB::add_to_batch(&mut batch, STORED_SHARDS_KEY, self.shards().clone());
+		}
+		// store everything in DB
+		self.db.write(batch)
+	}
+
+	/// purges a shard and its block from the db storage
+	pub fn purge_shard(&mut self, shard: &ShardIdentifierFor<SignedBlock>) -> Result<()> {
+		// Early return if shard is inexistent
+		if !self.shards.contains(shard) {
+			// check if DB is also empty
+			let db_shards = self.load_shards_from_db()?;
+			if !db_shards.contains(shard) {
+				warn!("Tried to purge already empty shard: {:?}", shard);
+				return Ok(())
+			}
+			warn!(
+				"Mismatch detected between inmemory shards and db shards, maybe try to reload DB? Guilty shard: {:?}", shard
+			);
+		}
+		let mut batch = WriteBatch::default();
+		// get last block of shard
+		let last_block = match self.last_blocks.get(shard) {
+			Some(last_block) => *last_block,
+			None => {
+				// Try to read from db:
+				self.load_last_block_from_db(shard)?
+					.ok_or_else(|| Error::LastBlockNotFound(format!("{:?}", *shard)))?
+			},
+		};
+		// remove last block from db storage
+		self.delete_last_block(&mut batch, &last_block, shard);
+
+		// Remove the rest of the blocks from the db
+		let mut current_block_number = last_block.number;
+		while let Some(previous_block) = self.get_previous_block(shard, current_block_number)? {
+			current_block_number = previous_block.number;
+			self.delete_block(&mut batch, &previous_block.hash, &current_block_number, shard);
+		}
+		// remove shard from list
+		// STORED_SHARDS_KEY -> Vec<(Shard)>
+		self.shards.retain(|&x| x != *shard);
+		// add updated shards to Batch DB
+		SidechainDB::add_to_batch(&mut batch, STORED_SHARDS_KEY, &self.shards);
+		// Update DB
+		self.db.write(batch)
+	}
+
+	/// purges a shard and its block from the db storage
+	/// FIXME: Add delete functions?
+	pub fn prune_shard_from_block_number(
+		&mut self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+		block_number: BlockNumber,
+	) -> Result<()> {
+		// Early return if shard is inexistent
+		if !self.shards.contains(shard) {
+			// check if DB is also empty
+			let db_shards = self.load_shards_from_db()?;
+			if !db_shards.contains(shard) {
+				warn!("Tried to prune already empty shard: {:?}", shard);
+				return Ok(())
+			}
+			warn!(
+				"Mismatch detected between inmemory shards and db shards, maybe try to reload DB? Guilty shard: {:?}", shard
+			);
+		}
+		// get last block of shard
+		let last_block = match self.last_blocks.get(shard) {
+			Some(last_block) => *last_block,
+			None => {
+				// Try to read from db:
+				self.load_last_block_from_db(shard)?
+					.ok_or_else(|| Error::LastBlockNotFound(format!("{:?}", *shard)))?
+			},
+		};
```

Comment:
all duplicate code of previous method.

Reply:
Extracted 		
```rust
// get last block of shard
		let last_block = match self.last_blocks.get(shard) {
			Some(last_block) => *last_block,
			None => {
				// Try to read from db:
				self.load_last_block_from_db(shard)?
					.ok_or_else(|| Error::LastBlockNotFound(format!("{:?}", *shard)))?
			},
		};
```
and removed obsolete shard check